File: box/find-kth-bit-in-nth-binary-string/log2.js
/**
 * @param {number} n
 * @param {number} k
 * @return {character}
 */
var findKthBit = function (n, k) {
  const begining = [null, 0, 1, 1, 1];
  if (k <= 4) return begining[k].toString();
  let simetry = 0;

  while (k > 4) {
    const logas = Math.log2(k);
    if (Number.isInteger(logas)) return (simetry ^ 1).toString();
    const prevCenter = 2 ** Math.floor(logas);
    k = 2 * prevCenter - k;
    simetry = 1 - simetry;
  }

  return (simetry ^ begining[k]).toString();
};

// console.log(findKthBit(3, 24));

let s = "";
for (let k = 1; k < 40; k++) {
  s += findKthBit(10, k);
}
console.log("js", s);

File: box/find-kth-bit-in-nth-binary-string/log2.py
import math
#
#
# This solution is based on the observation that the string is symetric
class Solution:
    def findKthBit(self, n: int, k: int) -> str:

        begining = [None, 0, 1, 1, 1]
        if k <= 4: return str(begining[k])  
        simetry = 0
        
        while k > 4:
            logas = math.log2(k)
            if logas.is_integer(): return str(simetry ^ 1)
            prev_center = round(2 ** math.floor(logas))
            k = 2 * prev_center - k
            simetry = 1 - simetry
        
        return str(simetry ^ begining[k])


# test cases to validate the solution
solution = Solution()

# print(solution.findKthBit(0, 24))

s = ""
for  m in range(1,40):
  s += str(solution.findKthBit(30, m))
print("ph", s)

            
        